### ISSUE TITLE
Can't move UI windows across screens

### ISSUE THREAD

**Issue by Colpocorto** [openMSX/openMSX#1623]
I am using a PC with two screens and sometimes happens that I can't move a new UI window from one screen to the other one. It just stops at some points (as if an invisible limit existed).

See this example video:

https://youtu.be/LGJDDi9_Dds

Version of OpenMSX: 19.1-909
OS: Windows 10 64bit Intel

**Comment by MBilderbeek** [comment]
@ocornut is this an issue on how we use Dear ImGUI?  If yes, any hint on what we're doing wrong?

**Comment by ocornut** [comment]
It seems to try keeping the window in the first screen for some reason.

can you open imgui demo>tools>metrics/debugger>viewports and report what’s listed under Monitors.?

**Comment by MBilderbeek** [comment]
@Colpocorto to do this in openMSX, go to Settings -> Advanced and scroll all the way down in the Advanced window. There click on the "imGUI Demo Window" (checkbox), which will open the window that @ocornut requested.

**Comment by Colpocorto** [comment]
> It seems to try keeping the window in the first screen for some reason.
> 
It allows moving the window partially, indeed.

> can you open imgui demo>tools>metrics/debugger>viewports and report what’s listed under Monitors.?

I'm flying out right now but will be back in a few days, I'll check it then.

Thank you for your help.

**Comment by MBilderbeek** [comment]
@Colpocorto any news on this?

**Comment by basalicante** [comment]
Hey guys, I was struggling with this too. But then I realized the second monitor has to be connected BEFORE starting imgui. If you do this the other way around, imgui doesn't detect the second monitor and you get the behavior that @Colpocorto described.

**Comment by MBilderbeek** [comment]
@Colpocorto can you please comment?

**Comment by Colpocorto** [comment]
My monitor is connected permanently.

**Comment by MBilderbeek** [comment]
Yes, but can you please check the request in https://github.com/openMSX/openMSX/issues/1623#issuecomment-2119358396 and https://github.com/openMSX/openMSX/issues/1623#issuecomment-2119364636 ?

**Comment by Colpocorto** [comment]
> Yes, but can you please check the request in [#1623 (comment)](https://github.com/openMSX/openMSX/issues/1623#issuecomment-2119358396) and [#1623 (comment)](https://github.com/openMSX/openMSX/issues/1623#issuecomment-2119364636) ?

Sorry, I missed that.
I'll try when it happens again (the issue doesn't occur always, only sometimes).

**Comment by MBilderbeek** [comment]
@Colpocorto has this never occurred again? If not, then please close the issue.

**Comment by Colpocorto** [comment]
@MBilderbeek I don't have access to a computer right now and won't have in a few more weeks.

Please feel free to close the ticket if you consider it shouldn't remain open.

**Comment by MBilderbeek** [comment]
It's not up for me to decide. I only have one monitor on my PC, so I can't test it myself. It's possible this has been fixed in the meantime though. But no need to hurry. 
If you think it can be closed because the problem doesn't occur anymore, please close the ticket. No problem if that takes a few more weeks or longer.

**Comment by Colpocorto** [comment]
I've found this problem again. The thing is that I've replaced one of the screens with a different problem but it has occurred again.

I think it might be caused by momentarily switching off one of the screens while the OpenMSX is still running. 

The panel is recognised again after switching on, but it doesn't allow to move the window entirely from one LCD to another (just like the video I sent when first opened the ticket).

**Comment by MBilderbeek** [comment]
@Colpocorto ok, so can you check Settings -> Advanced and scroll all the way down in the Advanced window. There click on the "imGUI Demo Window" (checkbox), which will open the window in which you can choose Tools>metrics/debugger>viewports and report what’s listed under Monitors?

**Comment by m9710797** [comment]
The "ImGui Demo Window" setting was moved to "main menu bar > Help > ImGui Demo Window".
It indeed contains a very extensive set of debug functionality (all implemented by the "Dear ImGui library" itself, not by openMSX).

**Comment by MBilderbeek** [comment]
Oh, right, I forgot that we moved it, sorry. Anyway @Colpocorto , ocornut requested that information so he can analyze this issue. Please provide it for your situation.

So, main menu bar > Help > ImGui Demo Window" > Tools > metrics/debugger > viewports  and then please report what's listed under Monitors.

**Comment by Colpocorto** [comment]
![image](https://github.com/user-attachments/assets/fd1712aa-4072-42f1-ac0e-1f5564bc7b6b)

**Comment by MBilderbeek** [comment]
@ocornut can you use this info that you requested to define a next step or draw a conclusion?

**Comment by ocornut** [comment]
Dear ImGui only knows about one monitor here.

Is this code properly compiled and running? 
```cpp
#if SDL_HAS_DISPLAY_EVENT
        case SDL_DISPLAYEVENT:
        {
            // 2.0.26 has SDL_DISPLAYEVENT_CONNECTED/SDL_DISPLAYEVENT_DISCONNECTED/SDL_DISPLAYEVENT_ORIENTATION,
            // so change of DPI/Scaling are not reflected in this event. (SDL3 has it)
            bd->WantUpdateMonitors = true;
            return true;
        }
#endif
```
https://github.com/openMSX/openMSX/blob/master/src/3rdparty/imgui/imgui_impl_sdl2.cc#L405C1-L413C7

Need SDL 2.0.26.

**Comment by Colpocorto** [comment]
> Dear ImGui only knows about one monitor here.
> 
However, the windows are actually drawn on the second monitor, but I doesn't allow to move the window entirely from one screen to the other.

**Comment by ocornut** [comment]
It doesn’t need to know about the monitors to draw there. It’s drawing in a window and the window can appear anywhere. But because we programmatically position windows we need to clamp them to reasonable boundaries which is why we need monitors.

**Comment by MBilderbeek** [comment]
> Dear ImGui only knows about one monitor here.
> 
> Is this code properly compiled and running?
> 
> ```c++
> #if SDL_HAS_DISPLAY_EVENT
>         case SDL_DISPLAYEVENT:
>         {
>             // 2.0.26 has SDL_DISPLAYEVENT_CONNECTED/SDL_DISPLAYEVENT_DISCONNECTED/SDL_DISPLAYEVENT_ORIENTATION,
>             // so change of DPI/Scaling are not reflected in this event. (SDL3 has it)
>             bd->WantUpdateMonitors = true;
>             return true;
>         }
> #endif
> ```
> 
> https://github.com/openMSX/openMSX/blob/master/src/3rdparty/imgui/imgui_impl_sdl2.cc#L405C1-L413C7
> 
> Need SDL 2.0.26.

As we're using SDL 2.30.1 that should be compiled in. (I also confirmed this on my Linux box. But I only have one monitor...)

**Comment by ocornut** [comment]
I guess you will need to debug why the code in ImGui_ImplSDL2_UpdateMonitors() is not doing its job, and verify that SDL_DISPLAYEVENT is triggered on the system when adding/removing a screen.

**Comment by MBilderbeek** [comment]
In the meantime I've updated to the latest SDL2 (2.30.7), but I doubt it will make a difference. But who knows...

**Comment by MBilderbeek** [comment]
@Colpocorto is this still happening with the latest openMSX development version?

**Comment by inchl** [comment]
same issue found on my box.
openmsx 20.0-9-g4bcb8d886, x86_64, release GL LASERDISC

"can you open imgui demo>tools>metrics/debugger>viewports and report what’s listed under Monitors.?"
see attachment

![Image](https://github.com/user-attachments/assets/ea3fc6ca-b1cd-44ff-bf4f-0b95f2ea563d)

**Comment by MBilderbeek** [comment]
> same issue found on my box. openmsx 20.0-9-g4bcb8d886, x86_64, release GL LASERDISC
> 
> "can you open imgui demo>tools>metrics/debugger>viewports and report what’s listed under Monitors.?" see attachment
> 
> ![Image](https://github.com/user-attachments/assets/ea3fc6ca-b1cd-44ff-bf4f-0b95f2ea563d)

Can you please try the following:
1. try with a newer version, the most recent development version is the most interesting to try
2. check whether you have switch off or plugged in a monitor whilst openMSX was running. I.e. can you also reproduce this if the other monitor was on and plugged in the whole openMSX session?
